cmd/thv-operator/controllers/virtualmcpserver_deployment.go

```diff
@@ -192,63 +193,248 @@ func (*VirtualMCPServerReconciler) buildEnvVarsForVmcp(
+		generatedSecretName := fmt.Sprintf("%s-oidc-client-secret", vmcp.Name)
+		env = append(env, corev1.EnvVar{
+			Name: "VMCP_OIDC_CLIENT_SECRET",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: generatedSecretName,
 					},
+					Key: "clientSecret",
 				},
-			})
+			},
+		})
+	}
+
+	return env
+}
+
+// buildOutgoingAuthEnvVars builds environment variables for outgoing auth secrets.
+func (r *VirtualMCPServerReconciler) buildOutgoingAuthEnvVars(
+	ctx context.Context,
+	vmcp *mcpv1alpha1.VirtualMCPServer,
+) []corev1.EnvVar {
+	var env []corev1.EnvVar
+
+	if vmcp.Spec.OutgoingAuth == nil {
+		return env
+	}
+
+	// Mount secrets from discovered ExternalAuthConfigs (discovered mode)
+	if vmcp.Spec.OutgoingAuth.Source == OutgoingAuthSourceDiscovered {
+		discoveredSecrets, err := r.discoverExternalAuthConfigSecrets(ctx, vmcp)
+		if err != nil {
+			ctxLogger := log.FromContext(ctx)
+			ctxLogger.V(1).Info("Failed to discover ExternalAuthConfig secrets, continuing without them",
+				"error", err)
+		} else {
+			env = append(env, discoveredSecrets...)
 		}
 	}
 
-	// Note: Other secrets (Redis passwords, service account credentials) may be added here in the future
-	// following the same pattern of mounting from Kubernetes Secrets as environment variables.
+	// Mount secrets from inline ExternalAuthConfigRefs
+	if vmcp.Spec.OutgoingAuth.Backends != nil {
+		inlineSecrets := r.discoverInlineExternalAuthConfigSecrets(ctx, vmcp)
+		env = append(env, inlineSecrets...)
+	}
 
-	return ctrlutil.EnsureRequiredEnvVars(ctx, env)
+	// Mount secret from Default ExternalAuthConfigRef
+	if vmcp.Spec.OutgoingAuth.Default != nil && vmcp.Spec.OutgoingAuth.Default.ExternalAuthConfigRef != nil {
+		defaultSecret, err := r.getExternalAuthConfigSecretEnvVar(
+			ctx, vmcp.Namespace, vmcp.Spec.OutgoingAuth.Default.ExternalAuthConfigRef.Name)
+		if err != nil {
+			ctxLogger := log.FromContext(ctx)
+			ctxLogger.V(1).Info("Failed to get Default ExternalAuthConfig secret, continuing without it",
+				"error", err)
+		} else if defaultSecret != nil {
+			env = append(env, *defaultSecret)
+		}
+	}
+
+	return env
+}
+
+// discoverExternalAuthConfigSecrets discovers ExternalAuthConfigs from MCPServers in the group
+// and returns environment variables for their client secrets. This is used for discovered mode.
+func (r *VirtualMCPServerReconciler) discoverExternalAuthConfigSecrets(
+	ctx context.Context,
+	vmcp *mcpv1alpha1.VirtualMCPServer,
+) ([]corev1.EnvVar, error) {
+	ctxLogger := log.FromContext(ctx)
+	var envVars []corev1.EnvVar
+	seenConfigs := make(map[string]bool) // Track which ExternalAuthConfigs we've already processed
+
+	// Get workload names from the group
+	workloadDiscoverer := workloads.NewK8SDiscovererWithClient(r.Client, vmcp.Namespace)
+	workloadNames, err := workloadDiscoverer.ListWorkloadsInGroup(ctx, vmcp.Spec.GroupRef.Name)
```

Comment:
This could absolutely be a follow up issue, but note that listing workloads in group at 2 different moments can give you 2 different sets (here and in ensureVmcpConfigConfigMap). I would prefer to do 1 List and pass the results around